napari/_qt/qt_main_window.py

```diff
@@ -1550,27 +1555,62 @@ def _screenshot(
         flash : bool
             Flag to indicate whether flash animation should be shown after
             the screenshot was captured.
-        size : tuple (int, int)
+        size : Optional[int, int]
             Size (resolution height x width) of the screenshot. By default, the currently displayed size.
             Only used if `canvas_only` is True.
-        scale : float
+        scale : Optional[float]
             Scale factor used to increase resolution of canvas for the screenshot.
             By default, the currently displayed resolution.
             Only used if `canvas_only` is True.
         canvas_only : bool
             If True, screenshot shows only the image display canvas, and
             if False include the napari viewer frame in the screenshot,
             By default, True.
+        fit_to_data_extent: bool
+            Tightly fit the canvas around the data to prevent margins from
+            showing in the screenshot. If False, a screenshot of the whole
+            currently visible canvas will be generated.
 
         Returns
         -------
         img : QImage
         """
         from napari._qt.utils import add_flash_animation
 
+        canvas = self._qt_viewer.canvas
+        prev_size = canvas.size
+        if fit_to_data_extent:
+            if not canvas_only:
+                raise ValueError(
+                    trans._(
+                        "'fit_to_data_extent' cannot be set to True if 'canvas_only' is"
+                        ' set to False',
+                        deferred=True,
+                    )
+                )
+            ndisplay = self._qt_viewer.viewer.dims.ndisplay
+            camera = self._qt_viewer.viewer.camera
+            old_center = camera.center
+            old_zoom = camera.zoom
+            if ndisplay > 2:
+                raise NotImplementedError(
+                    trans._(
+                        'fit_to_data_extent=True is not yet implemented for 3D. '
+                        'Please set fit_to_data_extent to False in 3D view.',
+                        deferred=True,
+                    )
+                )
+
+            self._qt_viewer.viewer.reset_view()
+            canvas.size = (
+                self._qt_viewer.viewer.layers.extent.world[1][
+                    -ndisplay:
+                ].astype(int)
+                + 1
+            )
+            self._qt_viewer.viewer.reset_view(margin=0)
+
         if canvas_only:
-            canvas = self._qt_viewer.canvas
-            prev_size = canvas.size
             if size is not None:
```

Comment:
As I understand, this branch collides with `fit_to_data_extent=True` without any warning?

Reply:
bump

Reply:
I mean that executing this branch with the provided size will reset the size to the provided one. 

So providing size and fit_to_data_extent=True should be mutually exclusive.

Reply:
@Czaki actually, I went to push a fix here, and it turns out that there is no public way for this to happen: only export_figure sets fit_to_data_extent (internally), while only screenshot can set size.

In order to add a mutually exclusive valueerror I would need to either have it be uncovered, or add a test for a private API, which I think is overly restrictive.

Since there is no way to trigger this error with a public API, I'm going to merge (pending tests) and make an issue about refactoring the logic in this internal function so that the flow is clearer. Actually I think a structural pattern matching call might be useful here, so we can expose valid paths (and those paths would be clear) and raise an error on all other combinations.